### DISCUSSION TITLE
Future refinement: PyScript shared environments - DISCUSS!

### DISCUSSION THREAD

**Discussion by ntoll** [pyscript/pyscript#1893] [Proposals]
Jotting these ideas down after @WebReflection and I had a chat where such ideas were proposed and initially explored. Our conclusion: ask folks for feedback for future refinement or rejection of the idea.

----

Currently we have no way for the Python terminal, Python editor or Python scripts to share the same environment.

A use case for why this is useful:

I am a teacher who has a Python script that does a bunch of important things. I want my students to be able to _explore the state of the universe once the script has finished executing_. The two ways of interactively working with PyScript are (as of time of writing) the terminal (`<script type="py" terminal worker>import code; code.interact()</script>`) and py-editor (`<script type="py-editor">...</script>`). I want my students to use one (or both?) of these methods to interact with the resulting state of the universe after the script has finished executing.

Questions:

* Is this actually a good idea? (Use case described above based on Chris at Tufts).
* Technically, how do such capabilities share a Python context? (Presumably, a shared env via a worker?)
* What gotchas? (E.g. if the `terminal` is blocking because it's waiting on `input` can you also run code from a `py-editor` given they're sharing the same worker context)

Ideas, constructive critique and discussion most welcome..!

**Comment by WebReflection** [comment]
Technically speaking:

  * I think being able to connect "*bricks*" **is** a good idea ... but ...
  * we need work on *polyscript* side to share the same environment ... dare I say that could be either out of the *main* thread or the *worker* one, and it should be transparent for the user as long as it's clear that **some** environment can run only in workers (terminal with a blocking input or editor which is by default in a worker)
  * the terminal story is a bit weird ... anyone can write as much Python code as they want **before** the `code.interact()` call is executed ... referring a terminal to an existent piece of code is also already possible: you have a `<pre id="terminal-dode">` tag, as example, with code in it, you can eval that in Python before asking to `code.interact()`
  * the editor that pollute the already known shared interpreter is indeed an interesting case, specially because we don't have a way to grab code from a foreign script **but** we could ... that could be a `source` attribute to me, so that an editor evaluate that code once before eventually evaluating its own code after ... this should fix all the things

In short, all I am saying is that there are ways to achieve the idea of a shared env already but of course this requires examples and documentation but it's possible, except for the `source` attribute in the editor that could, indeed, land in the next release, if we want to.

Any other approach requires a lot of tweaks in the *polyscript* module but *hell yeah* if I think those convoluted tweaks are cool and needed indeed!

**Reply by chrisbuerginrogers** [reply]
Hi all - as a fellow teacher (and a mechanical engineer - so unable to help on the low level issues) - I want to back up Nicholas's point - this was one of the real strengths with the older versions - this is the way my students would test out new ideas, debug old ones, and generally interact with their codes.  If there is any way to do this, it would be very helpful (I tried defining the two to be the same env - but that did nothing).  Thanks

**Reply by WebReflection** [reply]
the older version never worked in a non-blocking way (Web Workers) so everything is different now and, theoretically, for good (as it's, indeed, never blocking the rest of the main thread). If you don't use workers you have feature parity with what was possible but some new feature, or plugin, is landing on workers only ... hence the need to have clear requirements and best outcome we think of out of those requirements. I hope this makes somehow sense, yet thanks for your input, we definitively need to meet users expectations!

**Comment by JeffersGlass** [comment]
A related question came up organically in response to @Sadukie's talk at PyOhio 2023 - "If I define functions and data and stuff in my script, are they available in the [editor]?"

As you say Andrea, there's API/UX considerations here, but the theme of "I want to write code in a `<script>` tag and interact with it via an editor" is surely worth pursuing.

**Reply by ntoll** [reply]
> the theme of "I want to write code in a <script> tag and interact with it via an editor" is surely worth pursuing.

Quite. I don't think this is a case of if we do this... rather, how (given technical and UX limitations and considerations).

**Reply by MarcSkovMadsen** [reply]
+1 :-)
